### ISSUE TITLE
systemd units to auto-start scrcpy on connection

### ISSUE THREAD

**Issue by madduck** [Genymobile/scrcpy#3101]
Once #3085 is merged, the following files can be used to automatically fire up `scrcpy` for devices as they are connected.

Put the following into `$XDG_CONFIG_HOME/systemd/user/scrcpy@.service` for every user who wants `scrcpy` autostarted on connection:
```
[Unit]
Description=scrcpy client for device with serial %i
Documentation=man:scrcpy(1)
ConditionPathExists=%E/scrcpy/autostart/%i

[Service]
Environment="ARGS=--stay-awake --turn-screen-off --power-off-on-close --max-fps 10"
EnvironmentFile=-%E/scrcpy/autostart/DEFAULTS
EnvironmentFile=%E/scrcpy/autostart/%i
ExecStart=/usr/bin/scrcpy -s %i $ARGS
Restart=on-failure
RestartSec=2
SuccessExitStatus=2
Environment="DISPLAY=:0"
Environment="XAUTHORITY=%h/.Xauthority"
```
and reload the `systemd` instance: `systemctl --user daemon-reload`

Next, `touch` a file for each device serial number you want to support, e.g.:
```
touch $XDG_CONFIG_HOME/scrcpy/autostart/deadbeef01
```
This means that the `scrcpy@.service` unit will be able to fire up `scrcpy` for the device with the given serial number.

Now, plug in the device, and verify that it works:
```
systemctl --user start scrcpy@deadbeef01.service
```

If it does, finally place this into `/etc/udev/rules.d/local-scrcpy-autostart.rules`:
```
ACTION!="remove", SUBSYSTEM=="usb", ATTR{serial}=="?*", TAG+="systemd", \ 
  ENV{SYSTEMD_USER_WANTS}="scrcpy@$attr{serial}.service"
```

and run `udevadm control --reload`. Now unplug and replug the device, and `scrcpy` should start.

You can override the default arguments passed by setting `ARGS=…` in `$XDG_CONFIG_HOME/scrcpy/autostart/deadbeef01` for the specific device, or in `$XDG_CONFIG_HOME/scrcpy/autostart/DEFAULTS` for all devices that don't have specific configuration.

Released into the Public Domain. Please feel free to put on Wiki, or include in package, as you see fit.

**Comment by rom1v** [comment]
Thank you for the scripts :+1:

FYI, I wrote a tiny tool to execute shell commands when a new device is detected by adb: [`autoadb`](https://github.com/rom1v/autoadb).

**Comment by madduck** [comment]
@rom1v Yeah, I was aware of `autoadb`, but I think I concluded that since you're polling the adb port, this would be quite power-hungry of an implementation, and then someone told me about  `systemd-udevd`, and I took on the challenge.

**Comment by rom1v** [comment]
> I concluded that since you're polling the adb port

It's not polling, it performs a blocking read on localhost:5037. You could try with:

```bash
printf '0012host:track-devices' | nc localhost 5037
```

then connect/disconnect devices.

(But it was just for information. :wink:)

**Comment by Shelvak** [comment]
Hi, for simplicity I added a simple user systemd file and add the udev trigger like that, works like expected =) Would be great to add this to the readme =)

```bash
# /etc/udev/rules.d/98-scrcpy.rules
ACTION!="remove", ATTRS{idVendor}=="xxxx", ATTRS{idProduct}=="xxxx", TAG+="systemd", ENV{SYSTEMD_USER_WANTS}="scrcpy.service"
```

```bash
# ~/.config/systemd/user/scrcpy.service
[Unit]
Description=scrcpy autostart

[Service]
Type=simple
ExecStart=/usr/bin/scrcpy OPTS
Restart=on-failure
RestartSec=2
```

Hope this help =)
Cheers

**Comment by KAGEYAM4** [comment]
What' the update on this situation ? 

Which method to choose , the first one or the one previous one before me ?
